### PR TITLE
Update sharelatex to v4.1.1

### LINKED ISSUE #26: Update to Sharelatex 4.0
[ShareLatex 4.0.0](https://github.com/overleaf/overleaf/wiki/Release-Notes--4.x.x) has been released since 2023-05-30 and brings serveral features. Are there any updates planed?

I'd like to raise a PR if you can point out what needs to be changed.

## Release Brief

- A new [Source editor](https://www.overleaf.com/blog/were-retiring-our-legacy-source-editor) in addition to the Legacy editor will be available to users.
- Deleted projects and users can be automatically cleaned up after 90 days.
- TeXLive 2023 is now the default version for instances not running Sandboxed Compiles.
- The limit on a project’s editable content size (the sum of sizes of all editable files) has been increased from 5MB to 7MB.
- General performance and stability improvements to the application, along with many small improvements and bug fixes.

### PR DESCRIPTION
**Pull request by yzx9**
**tl; dr**:

- update sharelatex to v4.1.1
- format `docker-compose.*.yml`

## CHANGELOG

Sorry for the huge file changes in this PR, I will put some comments to make it easy to review.

For the `docker-compose.*.yml` file:
- Formatted it using two spaces for indentation (now a mix of four and two), and using double quotes for strings.
- Add a one-time image and command to setting up MongoDB as a replica set, see [Overleaf Release-Notes--4.x.x](https://github.com/overleaf/overleaf/wiki/Release-Notes--4.x.x#manually-setting-up-mongodb-as-a-replica-set) and [overleaf#1120](https://github.com/overleaf/overleaf/issues/1120)
- Pin `mongo:4.4` and `redis:6.2`

For `AuthenticationManager.js` and `ContactController.js`, we are fully sync from upstream and mark all changes with `>>>>>` and `<<<<<`

### Sharelatex 4.1

**Server Pro 4.1 is a security release for the application runtime.**

The Node.js runtime has been upgraded from version 16 to 18 ahead of the upcoming [deprecation of Node.js 16 on September 11, 2023](https://nodejs.org/en/blog/announcements/nodejs16-eol).

Only Server Pro 4.1 will operate with Node.js 18. All other supported versions of Server Pro require Node.js 16, which is being deprecated. We recommend that all customers upgrade to Server Pro 4.1 at their earliest convenience.

If an access point to your Server Pro instance is publicly accessible via the internet, such as a login page or redirect to it, it is particularly important that you upgrade to Server Pro version 4.1 before September 11, 2023.

### Sharelatex 4.0

- A new [Source editor](https://www.overleaf.com/blog/were-retiring-our-legacy-source-editor) in addition to the Legacy editor will be available to users.
- Deleted projects and users can be automatically cleaned up after 90 days.
- TeXLive 2023 is now the default version for instances not running Sandboxed Compiles.
- The limit on a project’s editable content size (the sum of sizes of all editable files) has been increased from 5MB to 7MB.
- General performance and stability improvements to the application, along with many small improvements and bug fixes.

---

Closes: #26 

Co-authored-by: @gizmo1-11

### CODE REVIEW

**Comment by yzx9** [comment]
While I was working on the changelog, I discovered that version 4.0.1 brought a breaking change. We had to upgrade from 3.5.10 to 4.0.1, but we only support 3.3.2 now, so #25 may be necessary. I'm not sure what we should do and I need your help @smhaller 

Official update instructions ([Release-Notes--4.x.x](https://github.com/overleaf/overleaf/wiki/Release-Notes--4.x.x#server-pro-401)):

> Important: Before upgrading to this new major version you will need to first upgrade your Overleaf Server Pro instance to version [3.5.10](https://github.com/overleaf/overleaf/wiki/Release-Notes-3.x.x#server-pro-3510) and migrate your projects to the new Full Project History system. Server Pro 4.0.0 will fail to start unless all the projects have been migrated.
>
> Read [Full Project History migration instructions](https://github.com/overleaf/overleaf/wiki/Full-Project-History-Migration).

**Comment by yzx9** [comment]
@smhaller Hi, any plan to merge it? I think it can be merged with an additional warning.

Althought bump backbone from 3.3.2 to 4.0.1 brings a BREAKING CHANGE, but users can upgrade their data by running the official 3.5.10 version image themselves. What we need to do (and can only do) is to add a warning.

**Comment by yzx9** [comment]
@smhaller any plan to merge it? I think it can be merged with an additional warning.

Althought bump backbone from 3.3.2 to 4.0.1 brings a BREAKING CHANGE, but users can upgrade their data by running the official 3.5.10 version image themselves. What we need to do (and can only do) is to add a warning.

**Comment by mooomooo** [comment]
Is it possible to also include oauth2 as an alternative to ldap auth?  The original from way long ago appears to have had both:
https://mirrors.sustech.edu.cn/git/sustech-cra/overleaf-ldap-oauth2/

Thanks for your work so far!

**Comment by yzx9** [comment]
> Is it possible to also include oauth2 as an alternative to ldap auth? The original from way long ago appears to have had both: https://mirrors.sustech.edu.cn/git/sustech-cra/overleaf-ldap-oauth2/
> 
> Thanks for your work so far!

@mooomooo Of course, I'm willing to work on it. I'll start once the issue is opened.

**Comment by mooomooo** [comment]
@yzx9 Thanks!  Issue #32 opened.

**Comment by smhaller** [comment]
> @smhaller Hi, any plan to merge it? I think it can be merged with an additional warning.
> 
> Althought bump backbone from 3.3.2 to 4.0.1 brings a BREAKING CHANGE, but users can upgrade their data by running the official 3.5.10 version image themselves. What we need to do (and can only do) is to add a warning.

Yes, I will merge it - in the upcoming week. Sorry for the delay. I will add a BREAKING Change warning (and I will try if just starting it once with the the sharelatex version 3.5.10 is enough to migrate).

File: docker-compose.traefik.yml
Comment (cl-parsons) [inline docker-compose.traefik.yml:1]: There seem to be an error in your indentation @yzx9 , the "command" block is too much unindented.

**Comment by VPaulV** [comment]
Hi! Just wanted to confirm it seems to work fine for me as well. Thank you for the implementation @yzx9
